From 4c40a7ae1b98d2b30c8f44c7e1447683da5c86fd Mon Sep 17 00:00:00 2001
From: Dong Aisheng <b29396@freescale.com>
Date: Thu, 9 Apr 2015 16:16:09 +0800
Subject: [PATCH 0392/1594] MLK-10629-2 mmc: sdhci-esdhc-imx: implement
 wifi_card_detect function

commit 97ab7ee61cf6d7dbaeaaa8338855c231a3b374ac from
git://git.freescale.com/imx/linux-2.6-imx.git

WiFi driver could call wifi_card_detect function to re-detect card,
this is required by some special WiFi cards like broadcom WiFi.
To use this function, a new property is introduced to indicate a wifi host.

Signed-off-by: Dong Aisheng <b29396@freescale.com>
---
 .../devicetree/bindings/mmc/fsl-imx-esdhc.txt      |    2 ++
 drivers/mmc/host/sdhci-esdhc-imx.c                 |   13 +++++++++++++
 2 files changed, 15 insertions(+), 0 deletions(-)

diff --git a/Documentation/devicetree/bindings/mmc/fsl-imx-esdhc.txt b/Documentation/devicetree/bindings/mmc/fsl-imx-esdhc.txt
index e478a2c..7bd878d 100644
--- a/Documentation/devicetree/bindings/mmc/fsl-imx-esdhc.txt
+++ b/Documentation/devicetree/bindings/mmc/fsl-imx-esdhc.txt
@@ -19,6 +19,8 @@ Optional properties:
   chapter, DLL (Delay Line) section in RM for details.
 - tuning-step : The delay cell steps in tuning procedure.
   Some boards need to set the tuning-step to make sure it can pass tuning procedure.
+- wifi-host : assigned as a wifi host.
+  This is required for some WiFi cards to do card detect
 - voltage-ranges : Specify the voltage range in case there are software
   transparent level shifters on the outputs of the controller. Two cells are
   required, first cell specifies minimum slot voltage (mV), second cell
diff --git a/drivers/mmc/host/sdhci-esdhc-imx.c b/drivers/mmc/host/sdhci-esdhc-imx.c
index dfa1265..5d5034b 100644
--- a/drivers/mmc/host/sdhci-esdhc-imx.c
+++ b/drivers/mmc/host/sdhci-esdhc-imx.c
@@ -136,6 +136,14 @@
 /* the IP supports eMMC HS400 */
 #define ESDHC_FLAG_SUP_HS400	BIT(9)
 
+static struct mmc_host *wifi_mmc_host;
+void wifi_card_detect(void)
+{
+	WARN_ON(!wifi_mmc_host);
+	mmc_detect_change(wifi_mmc_host, 0);
+}
+EXPORT_SYMBOL(wifi_card_detect);
+
 struct esdhc_soc_data {
 	u32 flags;
 };
@@ -996,6 +1004,11 @@ sdhci_esdhc_imx_probe_dt(struct platform_device *pdev,
 	if (of_find_property(np, "enable-sdio-wakeup", NULL))
 		host->mmc->pm_caps |= MMC_PM_WAKE_SDIO_IRQ;
 
+	if (of_get_property(np, "wifi-host", NULL)) {
+		wifi_mmc_host = host->mmc;
+		dev_info(mmc_dev(host->mmc), "assigned as wifi host\n");
+	}
+
 	mmc_of_parse_voltage(np, &host->ocr_mask);
 
 	/* call to generic mmc_of_parse to support additional capabilities */
-- 
1.7.5.4

